src/pybind/mgr/volumes/fs/operations/op_sm.py

```diff
@@ -10,15 +10,19 @@ class OpSm(object):
     FAILED_STATE = 'failed'
     FINAL_STATE  = 'complete'
     CANCEL_STATE = 'canceled'
+    SNAP_RETAINED_STATE = 'snap-retained'
```

Comment:
(providing an over-arching discussion for comments around state, incarnation and their values)

A subvolume with no incarnation (but existing snapshots), needs a manner to determine that it is in the `no-incarnation` state. This is needed as, some operations are allowed on these subvolumes and some others not.

Here are the operation classifications for a subvolume in the `no-incarnation` state,
- Allowed operations: create, snapshot ls, snapshot rm, snapshot protect, snapshot unprotect, snapshot clone
- Disallowed operations: rm, getpath, snapshot create, resize
- Operations being mused about: ls, info

The options to determine `no-incarnation` state are,
1) Leave the .meta file as is and determine incarnation is missing/deleted based on `path`, from the global section, not being found
- This is implicit, and needs a stat of the path and handling the same
- non-optimal and hence dropped

2) Change `path` in global section, to an empty string or remove `path` option from the global section
- more explicit, and does not need any further inspection of the subvolume

  I felt uncomfortable leaving the subvolume in the `complete` state in the .meta file, hence instead of the above,
3) Change `state` to denote a subvolume in the no incarnation state
- `path` value really does not matter, but again leaving it as is, can be misleading, so (2) needs to be done anyway

As long as we understand 'complete' state, as the subvolume is ready for certain operations, which includes it possibly being in `no-incarnation` state as well, we need not do (3) and can just stick to (2).

The addition of `incarnation` in the global section instead of dissecting the path to determine the same, was done as I felt this needed to be explicit than derived from path. This hence led to `EMPTY_INCARNATION` as well (or, removing the `incarnation` key from the .meta file). For v2 subvolumes, as long as the understanding is that the last component of the `path` *IS* the incarnation, we need not store the same explicitly.

If we agree to option (2) above for `no-incarnation` sub-state and that we can extract the `incarnation` from the `path` where it exists, the comments surrounding these changes can be addressed appropriately.